dev/benchmarking/ddl/1_tables.sql

```diff
@@ -0,0 +1,291 @@
+  IS 'The language the benchmark was written in (and presumably for).';
+COMMENT ON COLUMN public.benchmark_language.benchmark_language
+  IS 'The benchmark language. For example: Python';
+
+-- LANGUAGE_DEPENDENCIES
+CREATE TABLE IF NOT EXISTS public.language_dependencies
+(
+  id SERIAL
+  , benchmark_language_id integer NOT NULL
+  , dependencies jsonb NOT NULL DEFAULT '{}'::jsonb
+  , PRIMARY KEY (id, benchmark_language_id)
+  , FOREIGN KEY (benchmark_language_id) REFERENCES public.benchmark_language
+);
+COMMENT ON TABLE public.language_dependencies
+  IS E'@name language_dependencies.\n'
+      'A JSON string mapping dependencies to their versions for each language.';
+COMMENT ON COLUMN public.language_dependencies.dependencies
+  IS 'For example: ''{"boost": "1.69", "conda": "", "numpy": "1.15"}''.';
+
+-- LANGUAGE_VERSION
+CREATE TABLE IF NOT EXISTS public.language_version
+(
+  id SERIAL
+  , benchmark_language_id integer NOT NULL
+  , language_version citext NOT NULL DEFAULT ''
+  , PRIMARY KEY (id, benchmark_language_id)
+  , FOREIGN KEY (benchmark_language_id) REFERENCES public.benchmark_language
+);
+COMMENT ON TABLE public.language_version
+  IS 'The benchmark language version, e.g. ''2.7'' for Python 2.7';
+COMMENT ON COLUMN public.language_version.language_version
+  IS 'The version number used in the benchmark environment (e.g. ''2.7'').';
+
+-- BENCHMARK_TYPE
+CREATE TABLE IF NOT EXISTS public.benchmark_type
+(
+  id SERIAL PRIMARY KEY
+  , benchmark_type citext NOT NULL UNIQUE
+);
+COMMENT ON TABLE public.benchmark_type
+  IS 'The type of benchmark. For example "time", "mem", "peakmem", "track"';
+COMMENT ON COLUMN public.benchmark_type.benchmark_type
+  IS 'The type of units, so "time" for sec/ns/ms or "mem" for kB, MB, GB.';
+
+-- MACHINE
+CREATE TABLE IF NOT EXISTS public.machine
+(
+  id SERIAL PRIMARY KEY
+  , machine_name citext NOT NULL UNIQUE
+  , memory_bytes bigint NOT NULL
+  , cpu_actual_frequency_Hz bigint NOT NULL
+  , cpu_id integer NOT NULL
+  , gpu_id integer NOT NULL
+  , os_id integer NOT NULL
+  , FOREIGN KEY (cpu_id) REFERENCES public.cpu
+  , FOREIGN KEY (gpu_id) REFERENCES public.gpu
+  , FOREIGN KEY (os_id) REFERENCES public.os
+);
+COMMENT ON TABLE public.machine
+  IS 'Unique identifiers for a machine.';
+COMMENT ON COLUMN public.machine.machine_name
+  IS 'A unique machine name of your choice.';
+COMMENT ON COLUMN public.machine.memory_bytes
+  IS 'The output of `sysctl -n hw.memsize`.';
+COMMENT ON COLUMN public.machine.cpu_actual_frequency_Hz
+  IS 'The output of `sysctl -n hw.cpufrequency`.';
+
+-- UNIT
+CREATE TABLE IF NOT EXISTS public.unit
+(
+  id SERIAL PRIMARY KEY
+  , unit_name citext NOT NULL UNIQUE
+  , benchmark_type_id integer NOT NULL
+  , FOREIGN KEY (benchmark_type_id) REFERENCES public.benchmark_type(id)
+);
+COMMENT ON TABLE public.unit IS 'The actual units for a reported benchmark.';
+COMMENT ON COLUMN public.unit.unit_name
+  IS 'For example: ms, ns, microseconds.';
```

Comment:
we should probably have guidelines on whether to use abbreviations or not.

Reply:
+1; also I get confused by memory units and would like guidance on what to say for those (['kB' vs 'kiB' vs 'KB'](https://physics.nist.gov/cuu/Units/binary.html)). My personal opinion is unicode symbols are a pain (e.g. 'μ' or maybe even '\&mu;') and should be written out. Maybe this means everything should be written out.

Reply:
My preference would be on abbreviations. "k" is lowercase conventionally. "B" is byte. "kB" is 1000 bytes, "kiB" is 1024 bytes :-)

As for the special case of "µ", we can decide that people should use "u" instead. "us" is commonly used for microseconds.

Reply:
This could be pre-populated by admins, with insertion disabled for the "benchmark_type" and the "unit" table, to kind of force compliance with specific representations...